Query and Search blocks: support for Instant Search via `query_loop_block_query_vars` filter

michalczaplinski opened the pull request:
## What?
Implementation of Instant Search using the Search and Query Loop blocks. Added as a new experiment on the Gutenberg Experiments page.

Related to [#63053](https://github.com/WordPress/gutenberg/issues/63053)

https://github.com/user-attachments/assets/7ee8ddc7-472c-4575-a27e-af1972278c6c


## How does it work?

### Testing

1. Enable the `Instant Search and Query Block` Gutenberg experiment.
2. Insert a Query block anywhere on your site. 
3. Make sure to disable `Reload Full Page` checkbox in the Query block's Advanced Settings i.e. use "enhanced pagination" in that Query block.
4. Insert a Search block anywhere inside of that Query block.

When the Search block is inside of the Query block using "enhanced pagination", it automatically gets "updated" to an Instant Search block. 

Any search input in the Instant Search block gets passed as the `?instant-search-<queryId>=<search-term>` query search param in the URL.

### Multiple Query + Search blocks

It's possible to have multiple Instant Search blocks in a page/post/template. In such a case, ensuring their functionality is isolated is essential. For this to work, the `queryId` is part of the param syntax:
- `instant-search-<queryId>=<search-term>`

### Search button

The search block can contain a search button (in some configurations). However, when using the Instant Search functionality, the button is redundant because the list of posts updates as you type. For this reason, **in the editor**, when the Search block is placed inside the Query Loop block with enhanced pagination ON, the Block Controls related to the Search button are removed. The displayed name of the block (via `label`) is changed to `Instant Search (Search)`.

**On the frontend**, the Search button is also always removed for each (Instant) Search block.


https://github.com/user-attachments/assets/f75ba0a5-2672-4d36-839b-253f25fa4de8


### Pagination

The instant search functionality **respects** the pagination of the Query block, which uses the `query-<queryId>-page` syntax.

1. Every time the search is updated, the pagination of its respective query is reset back to page 1. 
2. The pagination numbers and next/previous, are updated to reflect the number of results in the search.
3. Clearing the search also resets the pagination back to page 1.

### Limitations

- Only the `Custom` query types are supported in the current experiment. The `Default` query types are currently not supported but will be in the future version. This is prototyped already in https://github.com/WordPress/gutenberg/pull/67289.

## Alternatives

Alternative approaches have been explored: 

- https://github.com/WordPress/gutenberg/pull/63147- modifying the server-rendering of the individual blocks (post-template, query-pagination, etc.)
- https://github.com/WordPress/gutenberg/pull/67013 - using the `render_block_context` filter.


## Further work

This is an initial prototype and the following features is **intentionally** not yet implemented:
- Support for the `Default` queries, including multiple `Default` queries on a single page/post/template. Those have been prototyped already in https://github.com/WordPress/gutenberg/pull/67289.
- Explore an alternative mechanism to convert the Search block into a Instant Search. Currently, this is done by nesting the Search inside of Query Loop with enhanced pagination ON, but a different, more explicit mechanism might be appropriate.
- Include contents of the Query No Results block but hide them with CSS ([comment](https://github.com/WordPress/gutenberg/pull/63147#discussion_r1860951647))

michalczaplinski committed:
add the filter and test

michalczaplinski committed:
remove

michalczaplinski committed:
Create instant search using `render_block_context` filter.

michalczaplinski committed:
Create instant search using `render_block_context` filter.

michalczaplinski committed:
Remove handling of inheritd query from `search/index.php`

michalczaplinski committed:
Handle case when query is defined in block context in DB.

michalczaplinski committed:
Guard against queryId being undefined in block context

michalczaplinski committed:
Fix the e2e test suite

michalczaplinski committed:
Move the filter to `/experimental` folder.

michalczaplinski committed:
Add an e2e test case if query.search attribute is present

michalczaplinski committed:
Remove the search button when instant search is enabled

michalczaplinski committed:
Use the query_loop_block_query_vars filter

michalczaplinski committed:
Do not delete pages and templates in e2e tests

michalczaplinski committed:
Set the pageId for Multiple Queries tests

michalczaplinski committed:
Fix the block name via metadata when Seach is instant.

michalczaplinski committed:
Remove stuff related to Default queries from `view.js`

michalczaplinski committed:
Add `attributes.metadata` to useEffect dependency array

michalczaplinski committed:
Remove `attributes.metadata` & label from dependency array

michalczaplinski committed:
Explain why we disable the react-hooks/exhaustive-deps lint

michalczaplinski committed:
Undo formatting changes

michalczaplinski committed:
Undo formatting changes

michalczaplinski committed:
Don't use `Promise.withResolvers()` in search block's view.js

michalczaplinski committed:
Remove the `render_block_context` filter from blocks.php

michalczaplinski commented:
This is ready for review now. Please note the updated PR description. 

I've closed https://github.com/WordPress/gutenberg/pull/67013 and https://github.com/WordPress/gutenberg/pull/63147 in favor of this PR as discussed in https://github.com/WordPress/gutenberg/issues/63053.

cc @luisherranz @gziolo @DAreRodz

fabiankaegy commented:
This is cool to see in core :) 

To me this highlights the need for a more customizable loading state though. By the nature of search being unpredictable we cannot really preload anything. So the amount of time it takes after each keystroke to get the results loaded more often than not is uncomfortably long. 

Yes we have the global navigation router loading bar. But that seems inappropriate here as it isn't a full page relead but instead is just updating this interactive region 🤔

This is totally not blocking here but I wonder if there is a way for us to make it much more obvious by either having a sort of spinner in the search bar itself or some overlayed spinner on the post list on a much shorter timer that appears when the router is fetching something.

fabiankaegy commented:
@luisherranz I just attempted to do this in a custom block of mine. Here are my findings:

1. It works but there is no way to hide the global animation bar whilst still getting the loading state
2. There is no way to determine the initiator of the current navigation which means you have to manually keep track of it

This here is how I did it:
```js
store('example/namespace', {
	context: {
		isActive: false,
	},
	state: {
		get isLoading() {
			const { state: routerState } = store('core/router');
			const { isActive } = getContext();

			return isActive && routerState.navigation.hasStarted;
		},
	},
	actions: {
		*search(event) {
			const context = getContext();
			const { actions } = yield import('@wordpress/interactivity-router');
			
			{...}

			context.isActive = true; // We need to somehow know that we are the reason for the loading state
			yield actions.navigate(url.toString());
			context.isActive = false; // and then reset it after
		},
	},
});
```

I think what could solve this if we added a new `isLoading` state to the router that we always set regardless of whether you disable the `loadingAnimation` option.  That way I could just use that prop and the global loading animation would get ignored.

luisherranz commented:
Thank you, Fabian. I've added it to the iteration of WordPress 6.8.

- https://github.com/WordPress/gutenberg/issues/67276

There's also this old discussion.

- https://github.com/WordPress/gutenberg/discussions/51924

To avoid blocking this work, for now we can use an internal state to display the spinner, like the one you show there.

michalczaplinski committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

michalczaplinski commented:
> This is totally not blocking here but I wonder if there is a way for us to make it much more obvious by either having a sort of spinner in the search bar itself or some overlayed spinner on the post list on a much shorter timer that appears when the router is fetching something.

@fabiankaegy Do you have an example of what the UI could look like for the loading state?

fabiankaegy commented:
@michalczaplinski I was thinking maybe starting with a simple inline spinner that gets shown absolutely positioned on the right of the search input next to the X (native clear button) 🤔

michalczaplinski committed:
Fix how directives are added to the form in the search block.

michalczaplinski committed:
Add e2e test for renaming Search block to "Instant Search" in the List View and Inspector Controls

- Also adds a new helper function to open the List View in the e2e test utils.

draganescu commented on lib/experimental/blocks.php:
```diff
@@ -118,6 +118,39 @@ function gutenberg_register_block_style( $block_name, $style_properties ) {
 }
 
+/**
+ * Adds the search query to Query Loop blocks if the instant search experiment is enabled.
```
```suggestion
 * Passes the search query param to Query Loop blocks, if the instant search experiment is enabled.
```

michalczaplinski replied:
Updated!

draganescu commented on packages/block-library/src/search/edit.js:
```diff
@@ -82,6 +83,26 @@ export default function SearchEdit( {
+	useEffect( () => {
+		if ( isEnhancedPagination ) {
+			// Add the name to the metadata
+			setAttributes( { metadata: { name: 'Instant Search' } } );
```
What does this do?

michalczaplinski replied:
It renames the block to "Instant Search". The updated name shows up in the ListView and the Inspector Controls.

We want to "upgrade" the Search block to Instant Search when 2 conditions are met:
1. The Search block is inside of a Query Loop block.
2. That Query Loop block has `enhancedPagination` enabled (The "Reload full page" switch is OFF)

gziolo replied:
Good catch, @draganescu. First, the user typically sets the `metadata.name`. What if the user set a different name for the block before enhanced pagination was activated? Second, the values visible in the UI should always be translatable. This attribute is also special, for example Pattern Overrides flow asks the user to provide the name for the block when activating the feature and the value gets stored in `metadata.name` and used as a reference for overriding the value. So side effects like this should be used following these considerations.

What is the goal of changing this name?

michalczaplinski replied:
My goal is to rename the block so that the user can distinguish between the Search block with the Instant Search and a regular Search block. 

You can see how it should look in the UI in the video in the PR description:

> For this reason, in the editor, when the Search block is placed inside the Query Loop block with enhanced pagination ON, the Block Controls related to the Search button are removed. The displayed name of the block (via label) is changed to Instant Search (Search).

https://github.com/user-attachments/assets/f75ba0a5-2672-4d36-839b-253f25fa4de8

> What if the user set a different name for the block before enhanced pagination was activated?

I didn't realize this was possible. Good catch. Given this, we probably shouldn't update the `metadata.name` then. 

We can just leave the block as is (no name change) or have an explicit opt-in. My thinking was that having the Search block just automatically become Instant Search when it's placed inside of a Query Loop with enhanced pagination is not explicit enough. Some users might miss that such a block is then an Instant Search block.

draganescu commented on packages/block-library/src/search/index.php:
```diff
@@ -90,6 +101,16 @@ function render_block_core_search( $attributes ) {
 			$input->set_attribute( 'tabindex', '-1' );
 		}
+
+		// Instant search is only available when using the enhanced pagination.
```
it'd be useful to say why?

michalczaplinski replied:
OK! I've updated the comment.

michalczaplinski committed:
Appease the PHP formatter

michalczaplinski committed:
Merge remote-tracking branch 'origin/trunk' into feature/search-query-using-vars-filter

draganescu commented on test/e2e/specs/interactivity/instant-search.spec.ts:
```diff
@@ -0,0 +1,650 @@
+            <!-- /wp:paragraph -->
+        <!-- /wp:query-no-results -->
+    </div>
+<!-- /wp:query -->`,
```
Nit pick but IMO using strings to set blocks is like using a snapshot which should not be the case in this context. We do the same in the zoom out context setting block content from strings which can change and invalidate the test for no reason.

michalczaplinski replied:
I'm not sure I understand. I think they are not like snapshots because if I change some whitespace in this string, it will still result in the same block structure in the editor. And it's that block structure that we're testing. Am I missing something?🙂

t-hamano replied:
If hard-coding HTML strings is not recommended, we can also use the following approach:

<details><summary>Details</summary>

```javascript
await admin.createNewPost();
await editor.insertBlock( {
	name: 'core/query',
	attributes: {
		enhancedPagination: true,
		queryId,
		query: {
			inherit: false,
			perPage: 2,
			order: 'desc',
			orderBy: 'date',
			offset: 0,
			search: 'Unique',
		},
	},
	innerBlocks: [
		{
			name: 'core/search',
			attributes: { label: '', buttonText: 'Search' },
		},
		{
			name: 'core/post-template',
			innerBlocks: [
				{
					name: 'core/post-title',
					attributes: { level: 3 },
				},
				{ name: 'core/post-excerpt' },
			],
		},
		{
			name: 'core/query-pagination',
			innerBlocks: [
				{ name: 'core/query-pagination-previous' },
				{ name: 'core/query-pagination-numbers' },
				{ name: 'core/query-pagination-next' },
			],
		},
		{
			name: 'core/query-no-results',
			innerBlocks: [
				{
					name: 'core/paragraph',
					attributes: {
						content: 'No results found.',
					},
				},
			],
		},
	],
} );
const id = await editor.publishPost();
```

</details> 

However, you need to remove [this code](https://github.com/WordPress/gutenberg/blob/a513013dfd6b691b933c3009dd69f707215b4557/test/e2e/specs/interactivity/instant-search.spec.ts#L136) so that the front-end page is not opened when the test starts.

michalczaplinski committed:
Update lib/experimental/blocks.php

Co-authored-by: Andrei Draganescu <andrei.draganescu@automattic.com>

michalczaplinski committed:
Update the comment in Search block to clarify when Instant Search is enabled

michalczaplinski committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

michalczaplinski committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

michalczaplinski committed:
Update e2e tests for Instant Search to verify block renaming in Inspector Controls and List View

gziolo commented on packages/block-library/src/search/edit.js:
```diff
@@ -82,6 +83,26 @@ export default function SearchEdit( {
 		style,
 	} = attributes;
 
+	const isEnhancedPagination = context?.enhancedPagination;
```
What about the expectations for the newly introduced logic in the case when the experiment is disabled?

michalczaplinski replied:
oh, I see. Yes, let's add a check here to make sure that the experiment is enabled. But first let's make sure we even want to change the name of the block in the UI or have a different way to differentiate the _Instant_ Search from regular Search 🙂 .

michalczaplinski replied:
I've discussed it with @jasmussen who suggested NOT updating the name of the block at all and instead adding a help text:

![Screenshot 2024-12-23 at 09 47 22](https://github.com/user-attachments/assets/907645a3-b038-41f6-b4c1-6a288f0e650e)

Hence, I'll remove the updates to `search/edit.js`

michalczaplinski replied:
Done in 15b80fe163c33cd457ed3f69a6f6eb9c4518c064

michalczaplinski replied:
ok, this is a little more complicated than I had thought because there is no dedicated prop that can be passed to e.g. `<InspectorControls>` to create this help text. It's generated from the `description` field of the block's `block.json`.

I'm looking into how we can add this secondary description in the least invasive way possible.

michalczaplinski replied:
I see two ways we can add this:

1. As a new `"secondaryDescription"` property in `block.json`. I don't really like that solution because it's pretty invasive for what seems like a small adjustment.
2. We could add a `<InspectorControls.Slot group="secondaryDescription">` in `block-inspector/index.js` right below the `<BlockCard>`: https://github.com/WordPress/gutenberg/blob/e5dca5461c0699b4ed902bbea375e40b0d2f803c/packages/block-editor/src/components/block-inspector/index.js#L250-L254

   This seems like a better option to me. The name of the group (`secondaryDescription`) can be adjusted of course 🙂.

michalczaplinski replied:
Another solution would be to add the `<InspectorControls.Slot group="secondaryDescription">` directly in the BlockCard component.

I've prototyped that approach in https://github.com/WordPress/gutenberg/pull/68438.

I'd appreciate thoughts on whether this is a sound approach!

michalczaplinski replied:
In the end, I've decided to (for now) remove any visual indicators that distinguish the Instant Search block from the regular Search block. 

We still might want to add the InspectorControls slot (https://github.com/WordPress/gutenberg/pull/68438) but that's not essential for the first version.

michalczaplinski replied:
Update: I've added a label (in the ListView only) following @t-hamano's tip.

See https://github.com/WordPress/gutenberg/pull/67181#issuecomment-2573474723

gziolo commented on packages/block-library/src/search/index.php:
```diff
@@ -29,11 +29,16 @@ function render_block_core_search( $attributes ) {
-	$show_button         = ( ! empty( $attributes['buttonPosition'] ) && 'no-button' === $attributes['buttonPosition'] ) ? false : true;
+	$input_id        = wp_unique_id( 'wp-block-search__input-' );
+	$classnames      = classnames_for_block_core_search( $attributes );
+	$show_label      = ( ! empty( $attributes['showLabel'] ) ) ? true : false;
+	$use_icon_button = ( ! empty( $attributes['buttonUseIcon'] ) ) ? true : false;
```
[`empty`](https://www.php.net/empty) returns a `bool` value. In effect, the ternary operator is redundant here.

michalczaplinski replied:
Good catch, thank you. 

Looks like I messed up the merge with trunk because this was changed in https://github.com/WordPress/gutenberg/pull/66857

michalczaplinski replied:
Fixed it in bdaaf2e3eaf604fd4747eaf5bf49c5d6d03d7a94

gziolo commented on packages/block-library/src/search/index.php:
```diff
@@ -29,11 +29,16 @@ function render_block_core_search( $attributes ) {
+	$show_label      = ( ! empty( $attributes['showLabel'] ) ) ? true : false;
+	$use_icon_button = ( ! empty( $attributes['buttonUseIcon'] ) ) ? true : false;
+	$show_button     = true;
+	if ( isset( $block->context['enhancedPagination'] ) && $block->context['enhancedPagination'] ) {
```
Should this check be limited to the situation where the Gutenberg experiment is enabled?

michalczaplinski replied:
ok, done in 97b3c845e11c0d1394949e55488a38a2d5eb1353

gziolo commented on packages/block-library/src/search/index.php:
```diff
@@ -90,6 +101,21 @@ function render_block_core_search( $attributes ) {
+		//
+		// Instant search functionality does not make sense without enhanced pagination
+		// because we might have to paginate the results of the search too!
+		if ( $enhanced_pagination ) {
+			wp_enqueue_script_module( '@wordpress/block-library/search/view' );
+
+			if ( $instant_search_enabled ) {
```
Can you elaborate here on what changes in the block when the enhanced pagination is enabled vs. when also the instant search is enabled? I don't understand the distinction.

michalczaplinski replied:
I think that the source of the confusion is that this could just be simplified to:

```php
if ( $enhanced_pagination && $instant_search_enabled) {
    wp_enqueue_script_module( '@wordpress/block-library/search/view' );
    
    $input->set_attribute( 'data-wp-bind--value', 'context.search' );
    $input->set_attribute( 'data-wp-on-async--input', 'actions.updateSearch' );
}
```

I don't think it makes sense to enqueue `@wordpress/block-library/search/view` if only enhanced pagination is enabled.

gziolo replied:
Right, that would help 👍🏻

michalczaplinski replied:
Updated it in bdaaf2e3eaf604fd4747eaf5bf49c5d6d03d7a94

DAreRodz commented on packages/block-library/src/search/view.js:
```diff
@@ -29,14 +32,25 @@ const { actions } = store(
 				const { isSearchInputVisible } = getContext();
 				return isSearchInputVisible ? '0' : '-1';
 			},
+			get isSearchInputVisible() {
+				const ctx = getContext();
+
+				// `ctx.isSearchInputVisible` is a client-side-only context value, so
+				// if it's not set, it means that it's an initial page load, so we need
+				// to return the value of `ctx.isSearchInputInitiallyVisible`.
+				if ( typeof ctx.isSearchInputVisible === 'undefined' ) {
+					return ctx.isSearchInputInitiallyVisible;
+				}
+				return ctx.isSearchInputVisible;
+			},
```
I see this derived state is only being used inside the `openSearchInput()` action, and `isSearchInputInitiallyVisible` is not read elsewhere, which makes me think that its function can be absorbed by `ctx.isSearchInputVisible`, right? 🤔 

Note that we changed how the state and context behave, and now, their data is not updated from the server during client-side navigation. I feel it would be safe to read the initial value from `ctx.isSearchInputVisible` directly.

michalczaplinski replied:
I think that you're right! Good catch 👍

michalczaplinski replied:
In addition to being redundant now that we've changed the behavior of state and context on client-side navigations, it's also not really necessary as this fix was needed when the Search button is visible. 

In this PR we're intentionally hiding the Search button (because such UX makes little sense for Instant Search) so this branch of the code should never be reached.

Done in 77e86d21693d72c0da94a30ef78b1bfff3b911af

DAreRodz reviewed (commented):
I see it's mostly there. 😄

Nothing to add apart from https://github.com/WordPress/gutenberg/pull/67181#discussion_r1891612057.

I share the same concerns @gziolo metioned about the code that should run only when the experiment is enabled. Also, the thing about the block name in the editor is tricky and, probably, there's a better way to let users know that the block is enhanced.

Once these issues are addressed, I guess the PR is ready to go.

michalczaplinski committed:
Simplify the logic for enqueuing `@wordpress/block-library/search/view`

michalczaplinski committed:
Put back the empty() calls that were erroneouslyreplaced with ternaries

michalczaplinski committed:
Only hide the search button if BOTH enhanced pagination & instant search experiment are enabled.

michalczaplinski committed:
Format PHP

michalczaplinski committed:
Removed redundant getter for `isSearchInputVisible` and simplified the logic to directly access the context.

michalczaplinski committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

michalczaplinski committed:
Check presence of instant search experiment in addition to the checking if enhancedPagination is enabled

michalczaplinski commented:
Instead of updating the block's name in the metadata when the Search block becomes an **Instant** Search block, I'm considering adding help text (more discussion in https://github.com/WordPress/gutenberg/pull/67181#discussion_r1899731026):

<img width="400" src="https://github.com/user-attachments/assets/907645a3-b038-41f6-b4c1-6a288f0e650e">

I opened a [PR to add a SlotFill to support this](https://github.com/WordPress/gutenberg/pull/68438) so this PR is currently blocked until we decide on an approach.

gziolo commented:
> I opened a https://github.com/WordPress/gutenberg/pull/68438 so this PR is currently blocked until we decide on an approach.

Is it essential to launch an experiment behind a flag?

michalczaplinski commented:
> Is it essential to launch an experiment behind a flag?

It's not essential 🙂. We can eliminate the indicator altogether for now and figure out a way to indicate that the block is enhanced before the block graduates from experimental status. Does that sound good?

gziolo commented on packages/block-library/src/search/index.php:
```diff
@@ -29,11 +29,27 @@ function render_block_core_search( $attributes ) {
+
+	// Check if the block is using the instant search experiment.
+	$gutenberg_experiments  = get_option( 'gutenberg-experiments' );
+	$instant_search_enabled = $gutenberg_experiments && array_key_exists( 'gutenberg-search-query-block', $gutenberg_experiments );
```
You can simplify all further checks, as instant search always requires the enhanced pagination:
```suggestion
	$instant_search_enabled = $enhanced_pagination && $gutenberg_experiments && array_key_exists( 'gutenberg-search-query-block', $gutenberg_experiments );
```

michalczaplinski replied:
Good point, I've updated it in dd42f507c9bc26c4544e35b1cdd9ee76ee919678

t-hamano commented:
> probably, there's a better way to let users know that the block is enhanced.

As an alternative approach to #68438, I imagine we could upe `__experimentalLabel` to let users know that instant search is enabled.

- Extend the `useBlockDisplayTitle` hook and `getBlockLabel` function to pass the `clientId` as a prop
- Add the text `(Instant search enabled)` after the block name if the Search block is in a Query Loop block and `enhancedPagination`  and `__experimentalEnableSearchQueryBlock` are enabled.

https://github.com/user-attachments/assets/a628f440-6f3a-495f-b4a2-c15f88b0d936

Here are the changes I made to this PR to achieve this:

<details><summary>Diff</summary>

```diff
diff --git a/packages/block-editor/src/components/block-title/use-block-display-title.js b/packages/block-editor/src/components/block-title/use-block-display-title.js
index a51b336554a..137c4ec1603 100644
--- a/packages/block-editor/src/components/block-title/use-block-display-title.js
+++ b/packages/block-editor/src/components/block-title/use-block-display-title.js
@@ -51,7 +51,12 @@ export default function useBlockDisplayTitle( {
                        }
 
                        const attributes = getBlockAttributes( clientId );
-                       const label = getBlockLabel( blockType, attributes, context );
+                       const label = getBlockLabel(
+                               blockType,
+                               attributes,
+                               context,
+                               clientId
+                       );
                        // If the label is defined we prioritize it over a possible block variation title match.
                        if ( label !== blockType.title ) {
                                return label;
diff --git a/packages/block-library/src/search/index.js b/packages/block-library/src/search/index.js
index 85770a23268..c81abebdf79 100644
--- a/packages/block-library/src/search/index.js
+++ b/packages/block-library/src/search/index.js
@@ -1,8 +1,10 @@
 /**
  * WordPress dependencies
  */
-import { __ } from '@wordpress/i18n';
+import { __, sprintf } from '@wordpress/i18n';
 import { search as icon } from '@wordpress/icons';
+import { select } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
 
 /**
  * Internal dependencies
@@ -18,6 +20,41 @@ export { metadata, name };
 
 export const settings = {
        icon,
+       __experimentalLabel( attributes, { clientId } ) {
+               const { label } = attributes;
+               const customName = attributes?.metadata?.name;
+
+               // Check if the block is inside a Query Loop block.
+               const queryLoopBlockIds = select(
+                       blockEditorStore
+               ).getBlockParentsByBlockName( clientId, 'core/query' );
+
+               // If the block is not inside a Query Loop block, return the block label.
+               if ( ! queryLoopBlockIds.length ) {
+                       return customName || label;
+               }
+
+               const queryLoopBlock = select( blockEditorStore ).getBlock(
+                       queryLoopBlockIds[ 0 ]
+               );
+
+               // Check if the Query Loop block has enhanced pagination enabled and
+               // if the `__experimentalEnableSearchQueryBlock` flag is enabled.
+               const hasInstantSearch = !! (
+                       queryLoopBlock?.attributes?.enhancedPagination &&
+                       window?.__experimentalEnableSearchQueryBlock
+               );
+
+               if ( ! hasInstantSearch ) {
+                       return customName || label;
+               }
+
+               return sprintf(
+                       /* translators: %s: The block label */
+                       __( '%s (Instant search enabled)' ),
+                       customName || label
+               );
+       },
        example: {
                attributes: { buttonText: __( 'Search' ), label: __( 'Search' ) },
                viewportWidth: 400,
diff --git a/packages/blocks/src/api/utils.js b/packages/blocks/src/api/utils.js
index 1a215036496..2f77bf19937 100644
--- a/packages/blocks/src/api/utils.js
+++ b/packages/blocks/src/api/utils.js
@@ -153,13 +153,18 @@ export function normalizeBlockType( blockTypeOrName ) {
  * @param {Object} blockType  The block type.
  * @param {Object} attributes The values of the block's attributes.
  * @param {Object} context    The intended use for the label.
- *
+ * @param          clientId
  * @return {string} The block label.
  */
-export function getBlockLabel( blockType, attributes, context = 'visual' ) {
+export function getBlockLabel(
+       blockType,
+       attributes,
+       context = 'visual',
+       clientId
+) {
        const { __experimentalLabel: getLabel, title } = blockType;
 
-       const label = getLabel && getLabel( attributes, { context } );
+       const label = getLabel && getLabel( attributes, { context, clientId } );
 
        if ( ! label ) {
                return title;
```

</details> 


One concern is that `__experimentalLabel` has no effect on the block card:

![block-title](https://github.com/user-attachments/assets/a1a79ef4-f766-42b5-9cd4-3aef4c96cc14)

michalczaplinski committed:
Simplify instant search and enhanced pagination checks.

michalczaplinski committed:
Update comments

michalczaplinski committed:
Refactor Instant Search tests to verify toolbar button visibility when Search block is inside Query block with enhanced pagination. Removed checks for renaming Search block to "Instant Search" in List View and Inspector Controls, focusing on button visibility instead.

michalczaplinski committed:
Update Search block label in ListView when it's an Instant Search

- Updated `getBlockLabel` to include `clientId` for improved label context.
- Modified `useBlockDisplayTitle` to utilize the updated `getBlockLabel` function.
- Introduced `__experimentalLabel` in the search settings to handle block labels based on query loop context and instant search status.

michalczaplinski committed:
Enhance Instant Search e2e tests to verify List View label updates

- Added a test to ensure the List View label updates correctly when the Search block is part of a Query block with enhanced pagination enabled.
- Verified that the label reflects "Instant search enabled" when applicable and reverts to "Search" when enhanced pagination is disabled.

michalczaplinski committed:
Update Search block label to default to 'Search' when no custom name is provided

michalczaplinski committed:
Remove default label attribute from Search block in Instant Search tests

- It's not needed because we fall back to 'Search' as last resort in `_experimentalLabel()` (previous commit).

michalczaplinski commented:
This works great @t-hamano, thank you! 👏 I'm not too worried about it not showing up in the Block Card for now. It's still better than not showing any updated label at all.

I've implemented your suggestion in 2db363aef52f776f1a405aab4d33013dc2747df9 and an e2e test for it in c20526fff7f5af97d4686c653a4d4a6726dbca36.

I also had to add a fallback to default to `"Search"` string in `__experimentalLabel()` in 6b975c4d413573e5e61f0d22b8b4ace4f0e7cdd9: 

   https://github.com/WordPress/gutenberg/blob/6b975c4d413573e5e61f0d22b8b4ace4f0e7cdd9/packages/block-library/src/search/index.js#L55

   Otherwise, if the block had no label, depending on the circumstances it ended up showing up as `undefined (Instant search enabled)` or just `(Instant search enabled)`:
   
   https://github.com/user-attachments/assets/4d16c213-7425-4ff9-be4d-7862c4f2dfde

t-hamano commented:
> Otherwise, if the block had no label, depending on the circumstances it ended up showing up as `undefined (Instant search enabled)` or just `(Instant search enabled)`:

Sorry, I think [the code I proposed](https://github.com/WordPress/gutenberg/pull/67181#issuecomment-2569108128) was not complete.

Originally in the List View, the display name and _label attribute_ were not synchronized. That means the List View only shows either "Search" or the custom name. Additionally, the text `Instant search enabled` may only need to be displayed in the List View.

So I think the logic should be like this:

```javascript
if ( context !== 'list-view' ) {
	return;
}

const blockTitle = __( 'Search' );

if ( ! queryLoopBlockIds.length ) {
	return customName || blockTitle;
}

if ( ! hasInstantSearch ) {
	return customName || blockTitle;
}

return sprintf(
	/* translators: %s: The block label */
	__( '%s (Instant search enabled)' ),
	customName || blockTitle
);
```

michalczaplinski committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

michalczaplinski committed:
Update __experimentalLabel to conditionally display block label in list view only. Adjusted fallback to use block title instead of label for better clarity.

michalczaplinski committed:
PHPCBF formatting

michalczaplinski commented:
@t-hamano I've updated it to use the logic you mentioned and re-tested 👍.

Would you mind taking another look?

t-hamano committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

t-hamano commented on packages/block-library/src/search/index.php:
```diff
@@ -27,11 +27,26 @@ function render_block_core_search( $attributes ) {
+	$enhanced_pagination = isset( $block->context['enhancedPagination'] ) && $block->context['enhancedPagination'];
+
+	// Check if the block is using the instant search experiment, which requires the enhanced pagination.
+	$gutenberg_experiments  = get_option( 'gutenberg-experiments' );
```
The `gutenberg-experiments` option is for the Gutenberg plugin, so I'm wondering if it's appropriate to write code like this in a file that is shipped to core. Of course, I think that once Instant Search is no longer experimental, this code can be removed.

t-hamano commented on packages/block-library/src/search/view.js:
```diff
@@ -71,6 +74,62 @@ const { actions } = store(
 					actions.closeSearchInput();
 				}
 			},
+			*updateSearch( e ) {
```
In CJK languages, an IME (Input Method Editor) is used to input complex characters. Therefore, for example, in a Japanese environment, the search results are updated even during IME composition. This is generally not desirable:

https://github.com/user-attachments/assets/33eff7ad-97ac-416b-bbd1-00094b492833


I guess we could address this in a follow-up, but ideally, we should ignore the event during composition and fire the updateSearch when the Enter key is pressed to commit the composition.

t-hamano replied:
This may be difficult to resolve at present, so let's consider it in a follow-up.

t-hamano commented on packages/e2e-test-utils-playwright/src/editor/open-list-view.ts:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @param this
+ */
+export async function openListView( this: Editor ) {
```
I don't think it's a good idea to add a new utility for E2E testing in this PR. If we want to open a list view, we should be able to simply use code like this:

```javascript
await page
	.getByRole( 'region', { name: 'Editor top bar' } )
	.getByRole( 'button', { name: 'Document Overview' } )
	.click();
```

t-hamano replied:
Fixd in 9fe9df540ed2bc12444ba1fe24ca46c5caadf297

t-hamano commented on test/e2e/specs/interactivity/instant-search.spec.ts:
```diff
@@ -0,0 +1,715 @@
+			).toISOString(),
+		} );
+
+		// Set the Blog pages show at most 2 posts
+		await requestUtils.updateSiteSettings( {
+			posts_per_page: 2,
+		} );
```
I think we need to restore the default settings via `afterAll`. Example:

https://github.com/WordPress/gutenberg/blob/a513013dfd6b691b933c3009dd69f707215b4557/test/e2e/specs/editor/plugins/post-type-templates.spec.js#L84-L92

t-hamano replied:
Fixed in 5ecf480e63be70260d616f8de3c556ef927c08f6

t-hamano commented on test/e2e/specs/interactivity/instant-search.spec.ts:
```diff
@@ -0,0 +1,715 @@
+			).toBeVisible();
+
+			// Type in search input and verify results update
+			await page.locator( 'input[type="search"]' ).fill( 'Unique' );
```
```suggestion
			await page.getByRole( 'searchbox' ).fill( 'Unique' );
```

Use accessible selectors whenever possible instead of `page.locator`.

t-hamano replied:
Fixed in 634e852779ab9a11396b34811de3ea5b8f174a9c

t-hamano reviewed (commented):
@michalczaplinski Sorry for the late reply!

I've resolved code conflicts and commented on things I've noticed throughout this PR to help move this PR forward. I'd be happy if you could take a look at them.

peterwilsoncc commented:
This looks really nice.

While testing, I noticed that there is an escaping bug for the search box when it contains quotes, they end up being escaped in the output `"Pens and pencils"` becomes `\"Pens and pencils\"` upon page refresh of an instant search URL. 

![escaped-quotes](https://github.com/user-attachments/assets/8f5f706c-c4a1-43d0-a82a-902e66fae73d)

If the instant search experiment is enabled, would it be possible to enable the feature on the search template with the option "Use instant search" for the search block? In that case it would be an inherited query so it's totally cool if the answer is no for the initial experiment.

t-hamano commented:
I would like to take over this PR.

t-hamano committed:
Merge branch 'trunk' into feature/search-query-using-vars-filter

t-hamano committed:
Remove custom playwright util

t-hamano committed:
e2e: Resttore original posts_per_page setting

t-hamano committed:
e2e: Use accesible selector

t-hamano committed:
Refactor all E2E tests

t-hamano committed:
Stabilize feature

t-hamano committed:
Remove unnecessary changes

t-hamano committed:
Stabilize feature

t-hamano committed:
Revert "Stabilize feature"

This reverts commit 88176a66d99ec69d7b889d88798b5e1a1b4ea25a.

t-hamano committed:
Revert "Stabilize feature"

This reverts commit 05938e0e87cf5ffd5e53fd0df32e5298ca4c92a6.

t-hamano committed:
Remove slashes from a URL string

t-hamano commented:
> If the instant search experiment is enabled, would it be possible to enable the feature on the search template with the option "Use instant search" for the search block? In that case it would be an inherited query so it's totally cool if the answer is no for the initial experiment.

@peterwilsoncc Instant Search is _automatically_ enabled when the following conditions are met:

- The "Instant Search and Query Block" experiment setting is enabled
- The "Reload full page" setting of the Query Loop block is disabled 
- The Search block is a direct child of the Query Loop block

Do you mean you want Instant Search to be optional rather than automatically enabled based on the above?

t-hamano commented on test/e2e/specs/interactivity/instant-search.spec.ts:
```diff
@@ -0,0 +1,655 @@
+				.pressSequentially( 'Test', { delay: 100 } );
+			await responsePromise;
+
+			// Check that only one request was made
+			expect( responseCount ).toBe( 1 );
```
This E2E test is failing, so we need to investigate why. It seems like a request is being sent for each character, even though a single request is expected:

```
Error: expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 4
```

t-hamano replied:
Fixed in 11ebed94ed545d661a64a0801b12308d10bc170f

t-hamano committed:
Debounce input event

t-hamano commented:
Ok, I think I'm ready to review again.

Having tackled this PR again, I have two suggestions:

- Stabilize instant search in this PR. This PR has sufficient e2e tests and does not make any breaking changes to the search block, so if there are any serious issues, we can always revert the feature.
- Make Instant Search optional. That is, add a toggle setting to the Search block so that users can disable Instant Search within a Query Loop block where "Reload full page" is disabled.
